### ISSUE TITLE
Disable IPv4/IPv6 addresses in SSH completion

### ISSUE THREAD

**Issue by 9ae8sdf76** [scop/bash-completion#427]
I've noticed when doing ssh completion it also includes any IPv4/IPv6 addresses that correlate to the hosts found. Is there an existing way to tell completion to disable ip addresses from completion? I see there is a `$BASH_COMPLETION_USER_FILE`, but I would not know how to remove entries. 

Or, would you all be agreeable to adding a `COMP_` environment variable to disable ip address translation?

Thank you!

**Comment by scop** [comment]
There kind of is, `_known_hosts_real` can be invoked with both `-4` and `-6` (each of those filters addresses of the respective family, and we use them when one completes e.g. "ssh -4 <kbd>Tab<kbd>". But there's nothing that would result in both being passed to it out of the box.

I don't feel that good about adding a new env var for controlling this (or pretty much anything for that matter), sorry.

I suppose one kind of easy but gross hack would be to make filtering both happen when one does e.g. "ssh -46 <kbd>Tab</kbd>", but that would be wrong, as we should actually not filter anything in that case.

**Comment by 9ae8sdf76** [comment]
Filtering using `-6` works for my purposes. Though I must admit I am a little disappointed about your stance against making an environment variable that'll do it behind the scenes. Having to remember to type `-6` each time feels like an extra burden (albeit small). It also changes the behavior of the underlying `ssh` command when all I need is to filter out IP addresses. 

Use Case: 
I work in a university and manage a growing list of servers for my department. For each new server added to my SSH config, there will be a corresponding IP address. Under normal usage - in my personal life, for example - this wouldn't be an issue. I could easily ignore the added noise. However, right now I have 22 servers to manage, and 44 entries due to the IP addresses. 

Please reconsider your stance. Not everything should be a convention. :)

**Comment by scop** [comment]
I stand firmly behind not adding configuration without thoroughly thinking through if we can make things just work. And even when we can't, it should be evaluated if the issue at hand is severe and unavoidable enough to warrant adding configuration for it. You asked for config at a phase where not everything had been covered, and I gave my gut feeling at that point (which still holds, and we still haven't covered everything).

Regarding your use case, here's what the terminal in front of me currently says:
> $ ssh <kbd>Tab</kbd>
> Display all 1605 possibilities? (y or n)^C

A lot of that can be thought of as noise if one looks at the complete list. But I practically never do for any actual real world use case I have. And I don't see the noise problem you describe, for me this works just fine. When I want to ssh somewhere, I have an idea where to. If it's a hostname, I type a few first chars of the hostname and the completion does not show any IP addresses. If it's an IPv4 address, I similarly type a few first and I'm not shown any hostnames and very rarely any IPv6 addresses get into the mix (this varies by the addresses one has around, sure). If I want to go to a IPv6 address, well, I confess that I practically never do. But I think the experience would be very similar to the one when I'm completing an IPv4 one.

**Comment by 9ae8sdf76** [comment]
> When I want to ssh somewhere, I have an idea where to.... If it's an IPv4 address, I similarly type a few first....

You seem like a pragmatic person, and from a practical standpoint why on earth would you choose an IP address over a hostname in a situation where the IP address is practically redundant? 

I don't know about the rest of the Internet, but I do not typically use IP addresses unless I absolutely have to, and even then it's usually because I need to set up the hostname to use instead. In other words, displaying IP addresses not only adds unnecessary noise to the output, but it does take time to resolve them through DNS or other name lookup mechanisms. 

Per your example of 1,605 possibilities, half of those are IP addresses that correspond to already existing DNS entries that completion already knows about. If for some reason you have IP addresses that do not resolve to hostnames, then more than likely you've defined an alias for it in your ssh config; I mean, why wouldn't you? Heck, I have a couple of very complex hostnames that I have no control over that I've aliased to simple names because it's easier to remember than `its5x5chb2wws` (that's an actual hostname).

I respect your pragmatism toward not overcomplicating an already complex software. But also, I feel that by choosing convention over configuration you're, respectfully, adding to that complexity instead of reducing it.

**Comment by scop** [comment]
I don't generally choose to use IP addresses if hostnames are available. Nowhere near half of my completions are IP addresses. But I have a bunch of entries that are only available as IP addresses.

**Comment by scop** [comment]
Anyway, additional filters can be added to completions offered by ssh with something like this in your shell startup files:
```bash
. /path/to/bash-completion/completions/ssh
complete -F _ssh -X '@(+([0-9.])|*:[a-f0-9]*)' ssh
```
It filters out things that look like IPv4 or IPv6 addresses from all ssh completions, not just hostnames, so it might hit something undesired. The globs could likely be made more strict matching to avoid that depending on one's scenario. Apply to other commands by finding out the completion using complete -p after invoking it, and add applicable  `-X ...` to them.

**Comment by Lekensteyn** [comment]
I agree with @scop not to add a configuration option for this, I think the added complexity of another option is not worth the little gain. What kind of hostnames do you have where the extra IP address completions cause issues?

I sometimes do complete IPv4 addresses, e.g. `192.168.x.x.`, but usally I type them out in full. Or use hostnames or aliases from ~/.ssh/config`. IP addresses are common in home networks where DHCP changes the address, but without making the hostnames available over DNS.

Some potential suggestions that do not involve modifying bash-completion:
- Avoid populating the known_hosts file with useless entries for temporary hosts (VMs and so on): `ssh -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no` (can also be done in `~/.ssh/config`).
- Type part of the hostname that is definitely not a valid IP (e.g. `s<tab>` -> serverXXX) :)

**Comment by scop** [comment]
On Sat, 11 Jul 2020 at 01:14, Peter Wu <notifications@github.com> wrote:

> ssh -oUserKnownHostsFile=/dev/null -oStrictHostKeyChecking=no


This is something I do every now and then. If specified with a space
between `-o` and the option name, it completes nicely too, even without
having to match case. (I wonder if we should support completing those
without the mentioned space, too.)
